lib/Dialect/OM/Transforms/FreezePaths.cpp

```diff
@@ -0,0 +1,240 @@
+//===- FreezePaths.cpp - Freeze Paths pass ----------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Contains the definitions of the OM freeze paths pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/HW/HWInstanceGraph.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/HW/InnerSymbolTable.h"
+#include "circt/Dialect/OM/OMAttributes.h"
+#include "circt/Dialect/OM/OMOps.h"
+#include "circt/Dialect/OM/OMPasses.h"
+
+using namespace circt;
+using namespace om;
+
+namespace {
+struct PathVisitor {
+  PathVisitor(hw::InstanceGraph &instanceGraph, hw::InnerRefNamespace &irn)
+      : instanceGraph(instanceGraph), irn(irn) {}
+  LogicalResult processPath(PathOp path);
+  LogicalResult run(Operation *op);
+  hw::InstanceGraph &instanceGraph;
+  hw::InnerRefNamespace &irn;
+};
+} // namespace
+
+static LogicalResult getAccessPath(Value value, size_t fieldId,
```

Comment:
Nice!

"One day" it'd be nice for ExportVerilog to be able to handle this in verbatim substitutions too! Not for this, just woohoo! :+1: